### ISSUE TITLE
RFC: A better data structure for `Headers`

### ISSUE THREAD

**Issue by armanbilge** [http4s/http4s#6720]
Spinning out of discussion in https://github.com/http4s/http4s/pull/6201. cc @wjoel @sideeffffect @CremboC

Off the top of my head there are a few issues at play I believe:
1. A better data structure than `List` for appending new headers.
2. A better data structure than `List` for looking up headers.
3. A data structure that doesn't preserve order, since it's not required by spec and other frameworks don't.

**Comment by armanbilge** [comment]
And also: is it _really_ a breaking change if we stop preserving header order in 0.23.x.

**Comment by bplommer** [comment]
Would `Map[CIString, NonEmptySet[String]]` work?

**Comment by bplommer** [comment]
> And also: is it really a breaking change if we stop preserving header order in 0.23.x.

In principle no. In practice it will break enough app tests to deter people from updating, which in turn will leave people's apps with unpatched known vulnerabilities.

**Comment by armanbilge** [comment]
Not quite. Ordering of repeated headers is important (although overall ordering of the headers themselves is not).

**Comment by bplommer** [comment]
So `Map[CIString, NonEmptyChain[String]]`?

**Comment by sideeffffect** [comment]
What would be the pros and cons of `NonEmptyChain` vs `NonEmptyVector`?

**Comment by armanbilge** [comment]
@sideeffffect https://github.com/typelevel/cats/pull/4264#issuecomment-1172973562
